segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -392,66 +401,213 @@ private boolean isFatalException(Throwable ex) {
+                    // Only record the commit if we had no failures.
+                    this.checkpointPolicy.recordCommit(commitArgs.getDataFrameLength());
+                }
             }
-
-            this.logUpdater.flush();
-            this.checkpointPolicy.recordCommit(commitArgs.getDataFrameLength());
         }
 
         /**
          * Rolls back all pending Metadata changes and fails all pending operations.
          *
          * @param ex The cause of the failure. The operations will be failed with this as a cause.
+         * @param commitArgs The Data Frame Commit Args that triggered this action.
+         */
+        void fail(Throwable ex, DataFrameBuilder.CommitArgs commitArgs) {
+            Map<CompletableOperation, Throwable> toFail = new HashMap<>();
+            try {
+                synchronized (this.lock) {
+                    collectFailureCandidates(ex, commitArgs, toFail);
+                }
+            } finally {
+                toFail.forEach(this::failOperation);
+                autoCompleteIfNeeded();
+            }
+
+            if (isFatalException(ex)) {
+                synchronized (this.lock) {
+                    this.stopException = ex;
+                }
+
+                CallbackHelpers.invokeSafely(this.fatalExceptionCallback, ex, null);
+            }
+        }
+
+        /**
+         * Fails the given Operation either with the given failure cause, or with the general stop exception.
+         *
+         * @param operation    The CompletableOperation to fail.
+         * @param failureCause The original failure cause. The operation will be failed with this exception, unless
+         *                     the general stopException is set, in which case that takes precedence.
+         */
+        void failOperation(CompletableOperation operation, Throwable failureCause) {
+            synchronized (this.lock) {
+                if (this.stopException != null) {
+                    failureCause = this.stopException;
+                }
+            }
+
+            operation.fail(failureCause);
+        }
+
+        /**
+         * Rolls back any metadata that is affected by a failure for the given commit args and collects all CompletableOperations
+         * off the pendingOperations queue that are affected. While the metadata is rolled back, the operations themselves
+         * are not failed (since this method executes while holding the lock).
+         *
+         * @param ex         The Exception that triggered the failure.
+         * @param commitArgs The CommitArgs that points to the DataFrame which failed to commit.
+         * @param target     A Map where the failure candidates will be collected. Key=Operation, Value=Failure Exception.
          */
-        void fail(Throwable ex) {
+        @GuardedBy("lock")
+        private void collectFailureCandidates(Throwable ex, DataFrameBuilder.CommitArgs commitArgs, Map<CompletableOperation, Throwable> target) {
```

Comment:
There are two invocations of this method. It does what you described in one case. In the other case, I have an existing map to which I add in 3 places. So I  need this passed as an argument.